Add "ORDER BY" possibility for QgsFeatureRequest

m-kuhn opened the pull request:
```
layer.getFeatures(QgsFeatureRequest().addOrderBy('"z"')) # will order by the field z
```

ORDER BY will be sent to provider for spatialite and postgres

m-kuhn committed:
More use of implicit sharing of QgsExpression

m-kuhn committed:
Add Order By Clause

m-kuhn committed:
Compile order by for postgres

m-kuhn committed:
Add order by test

m-kuhn committed:
Don't double sort in vector layer.

nyalldawson commented on src/core/qgsfeaturerequest.h:
```diff
@@ -85,6 +85,64 @@ class CORE_EXPORT QgsFeatureRequest
+         *
+         * @param expression The expression to use for ordering
+         * @param ascending  If the order should be ascending (1,2,3) or descending (3,2,1)
+         *                   If thr order is ascending, by default nulls are last
```
thr ->the

nyalldawson commented on src/core/qgsfeaturerequest.h:
```diff
@@ -85,6 +85,64 @@ class CORE_EXPORT QgsFeatureRequest
 
     /**
+     * @brief The OrderByClause class represents an order by clause for a QgsFeatureRequest
+     */
```
should have an "added in note"

nyalldawson commented on src/core/qgsfeaturerequest.h:
```diff
@@ -222,9 +304,11 @@ class CORE_EXPORT QgsFeatureRequest
     bool acceptFeature( const QgsFeature& feature );
 
+    void setOrderBy( const QList<OrderByClause>& orderBy );
+    QList<OrderByClause> orderBy() const;
```
what's this for? isn't it the same as `orderBys()`?

nyalldawson commented on src/core/qgsfeatureiterator.cpp:
```diff
@@ -18,6 +18,83 @@
+              return v1.toString().localeAwareCompare( v2.toString() ) < 0;
+            else
+              return v1.toString().localeAwareCompare( v2.toString() ) > 0;
+        }
```
do bools need to be handled here?

m-kuhn replied:
done

nyalldawson commented on src/providers/postgres/qgspostgresfeatureiterator.cpp:
```diff
@@ -111,12 +112,34 @@ QgsPostgresFeatureIterator::QgsPostgresFeatureIterator( QgsPostgresFeatureSource
+    }
+    else
+    {
+      mOrderByCompiled = false;
```
should this also clear orderByParts? is it an issue if a partial order by clause is passed, or is that desirable?

m-kuhn replied:
Not sure.

Partial compilation for sure won't help.

Having the first of two clauses pass may help to presort and reduce the time the local sort fallback requires (but the expensive expression evaluation will still need to take place)

I think it should stop at the first failing clause. This way it will only use partially compiled important clauses and not pre-sort on unimportant ones.

nyalldawson commented on tests/src/python/providertestbase.py:
```diff
@@ -120,6 +120,31 @@ def testGetFeaturesCompiled(self):
+
+        request = QgsFeatureRequest().addOrderBy('name', False, False)
+        values = [f['name'] for f in self.provider.getFeatures(request)]
+        self.assertEquals(values, ['Pear', 'Orange', 'Honey', 'Apple', NULL])
```
Can you add some tests which combine orderBy with limit to ensure that the order by is correctly applied before the limits?

nyalldawson replied:
I also think there should be a test for ordering by an expression, and for ordering by multiple columns/expressions

m-kuhn replied:
Will do

nyalldawson replied:
Oh - one more test which would be good to see would be a test which relies on case-sensitivity of sorting. This would help identify if any providers have different behaviour with case sensitive sorting when compared to the reference QGIS implementation.

m-kuhn replied:
Already done ;)

nyalldawson replied:
Where? I'm meaning a test where a field contains values which are the same apart from capitalization ( eg "aA", "aa", "Aa", "a a", "a"), so that we can test that they are always sorted the same

m-kuhn replied:
It's not yet pushed. Fixing some of the issues which surfaced while implementing more tests currently.

I'm just ordering by name2 (that should already show if the order priority is alphabetical or captialization)

m-kuhn replied:
And exactly 2.1 seconds after writing this I realized that spatialite actually does trigger this...

m-kuhn replied:
Pushed the latest version (failing spatialite)

m-kuhn replied:
Pushed again, working spatialite

nyalldawson commented on src/core/qgsfeaturerequest.h:
```diff
@@ -175,6 +233,30 @@ class CORE_EXPORT QgsFeatureRequest
+     *                   If thr order is descending, by default nulls are first
+     */
+
+    QgsFeatureRequest& addOrderBy( const QString &expression, bool ascending = true );
```
might be worth noting in the docs that calling this multiple times will add additional order by clauses, not remove the existing ones

nyalldawson commented:
LGTM with the comments noted above. Great stuff! 

Any plans to implement for other providers?

m-kuhn commented:
Should be easy enough to implement :)
I think I'll do for spatialite and leave the other DB systems to people with possibilities to test.

nirvn commented:
@m-kuhn wouhou, spatialite! :)

m-kuhn committed:
Compile order by for Spatialite

m-kuhn commented:
Thanks a lot for your review @nyalldawson 
Found some additional issues which should be fixed.

Waiting for coverage and postgres test results from travis

SebDieBln commented:
Note that sort order is not provider specific, at least not for Postgres:
[_The collation feature allows specifying the sort order and character classification behavior of data per-column, or even per-operation._](http://www.postgresql.org/docs/current/static/collation.html)

SebDieBln commented:
Also for sqlite/spatialite:
https://www.sqlite.org/datatype3.html#collation

m-kuhn commented:
We can add an additional option to control each order by clause if it's required. But I'd say a "best effort for equal behavior" in some standard situations should be sufficient to merge this PR.

I.e. I believe that we will never ever be able to provide fully compatible behavior in every situation for every provider. Let's try our best with the available resources.

The latest commit adds a check if the expression compiler is enabled so paranoid users can turn it off to have a stable but slow behavior if preferred.

m-kuhn committed:
Add compiler setting check to order by compilation

nyalldawson commented on tests/src/python/providertestbase.py:
```diff
@@ -120,6 +120,51 @@ def testGetFeaturesCompiled(self):
+        # Case sensitivity
+        request = QgsFeatureRequest().addOrderBy('name2')
+        values = [f['name2'] for f in self.provider.getFeatures(request)]
+        self.assertEquals(values, ['Apple', 'Honey', 'NuLl', 'oranGe', 'PEaR'])
```
Is this correct? I'd have expected 'oranGe' to come first.

m-kuhn replied:
It's what QString::localeAwareCompare does by default.

But I start to think that this should not be local aware and rather use compare instead.

m-kuhn replied:
That's what QString::compare does:

    [u'Apple', u'Honey', u'NuLl', u'PEaR', u'oranGe']

m-kuhn replied:
Investigating the issue:

http://www.postgresql.org/docs/9.1/static/locale.html

> The locale settings influence the following SQL features:
Sort order in queries using ORDER BY or the standard comparison operators on textual data

So it seems not even postgres has a portable and stable behavior across systems internally.

nyalldawson replied:
Bah...!

My thoughts... Either:
- just make a note of this in the docs, something along the lines of "Note that behavior of case sensitivity while sorting depends on the platform, user's locale and provider support."
- add extra flags for "ignore case during sorting" , "lowercase before uppercase"

I'm wondering now if it's more future proof to make the ascending and null first boolean arguments to OrderByClause instead use flags rather than seperate arguments. Thoughts?

m-kuhn committed:
Run travis tests also with build warnings.

m-kuhn committed:
cond

m-kuhn committed:
Fix csv test dataset

m-kuhn committed:
Remove .cpp files from doxymentation

Proposes to remove .cpp files from the doxymentation. They are not
considered to be part of the public API, so no reason to scan them for
symbols which are not exported anyway.

m-kuhn commented:
Merged in d96a2748b67a29dfde7 for wider testing.

Added doxymentation for the order by behavior for strings and other implementation details.

Please add more comments here if there are still problems.
